core/include/scipp/core/variable.tcc

```diff
@@ -613,6 +614,52 @@ Variable::Variable(const units::Unit unit, const Dimensions &dimensions,
                                                     std::move(values_),
                                                     std::move(variances_))) {}
 
+template <class T, class... T1, class... T2>
+Variable Variable::createVariable(units::Unit &&u, Dims &&d, Shape &&s,
+                                  std::tuple<T1...> &&val,
+                                  std::tuple<T2...> &&var) {
+  if constexpr (std::is_constructible_v<Vector<T>, T1 &&...> &&
+                std::is_constructible_v<Vector<T>, T2 &&...>) {
+    auto values = detail::make_move_from_tuple<Vector<T>>(val);
+    auto variances = detail::make_move_from_tuple<Vector<T>>(var);
+    constexpr bool has_val = (sizeof...(T1) > 0);
+    constexpr bool has_var = (sizeof...(T2) > 0);
+    auto dms = Dimensions{d.dims, s.shape};
+    if constexpr (has_val && has_var)
+      return Variable(u, dms, std::move(values), std::move(variances));
+    if constexpr (has_val)
+      return Variable(u, dms, std::move(values));
+    if constexpr (has_var)
+      throw std::logic_error("Can't have variance without values");
+    else {
+      auto res = makeVariable<T>(dms);
+      res.setUnit(u);
+      return res;
+    }
+  } else
+    throw except::TypeError(
```

Comment:
Can we move this type of fall back to the `CallDtype`: put the fall back functions, which trows, if the `dtype` is incompatible with arguments. In this particular case we can't construct  `Variable` with type `float` from `Values` contains `string`.